### ISSUE TITLE
How to generate in-distribution embeddings?

### ISSUE THREAD

**Issue by liuzh-buaa** [HKU-MedAI/bnn_uncertainty#1]
Thank you for your excellent work.

In the paper Section 3.2, "First, we sample $s$ weights for every data point in the training data, and generate representations for each data point using the sampled weights. In total, $n_1$ representations are generated from the training data."

As we sample $s$ weights, for each training data, we can obtain $s$ representations. I am confused about how to deal with them. Specifically, if we assume the size of training data is $n$, I wonder
(1) whether $n_1=n \times s$. That is to mix all representations of all training data, and then calculate $\bar{X}_1, \Sigma_1$.
(2) Or whether $n_1=n$. That is to mean $s$ representations of each data first, then to mix $n$ representations to calculate $\bar{X}_1, \Sigma_1$.

**Comment by howardchanth** [comment]
Hi thank you for your appreciation of our work.

I think there will be some information loss if we adopt (2) and perform a mean reduction first on each data point. Hence in the paper we adopt (1) to generate a more representative posterior predictive distribution of the whole training data.

**Comment by liuzh-buaa** [comment]
Thank you for your explanation. 

I have another question. The process involves first training a BNN, and then fixing this BNN to generate embeddings. I'm curious to know which layer's output of the BNN generates these embeddings. 

For example, for image classification task, the final output of the BNN would be the probabilities after the softmax layer. In this case, are the outputs of the final layer before the softmax layer as the embeddings?

**Comment by howardchanth** [comment]
Yes, this is correct. We treat the outputs of the last layer of the BNN encoder as the embeddings. Then the embeddings can be input into a classification layer (e.g., a Bayesian linear layer followed by a softmax) or a prediction layer for other downstream tasks (e.g., regression).
